index.js

```diff
@@ -0,0 +1,254 @@
+ * file) and all files in its dependency tree.
+ *
+ * As optimization of chunks occur, modules are moved out of entry chunks and
+ * into new chunks that this plugin creates.  The plugin also updates
+ * dependencies between chunks (via the 'parents' property on chunks) based on
+ * whether a module already exists in shared chunk or not.
+ *
+ * Shared chunks are created in the order they are specified.
+ *
+ * A set of shared chunks can be specified in a webpack.config.js like so:
+ *
+ *  plugins: [
+ *      new SharedChunkPlugin({
+ *          specs: [
+ *              {
+ *                  name: "vendor",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /vendor/.test(module.resource),
+ *              },
+ *              {
+ *                  name: "components",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /components/.test(module.resource),
+ *              },
+ *              {
+ *                  name: "feature-shared",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /features/.test(module.resource),
+ *                  minChunks: 2,
+ *              },
+ *          ]
+ *      })
+ *  ],
+ *
+ * Each of the 'specs' can contain the following options:
+ * - name: the name of the chunk being generated
+ * - selectedChunks: an array of entry chunks to consider when looking for
+ *                   common modules
+ * - moduleFilter: a predicate which can be used to determine whether a module
+ *                 should be included in the shared chunk that is currently
+ *                 being created
+ * - minChunks (optional): a number between 1 and Infinity that specifies how
+ *                         many entry chunks it must appear in before it is
+ *                         extracted to the shared chunk.
+ */
+class SharedChunkPlugin {
+    constructor(options) {
+        this.specs = options.specs;
+    }
+
+    apply(compiler) {
+        compiler.plugin("compilation", (compilation) => {
+            compilation.plugin(["optimize-chunks"], (allChunks) => {
+                const globalModulesSet = new Set();
+                for (const spec of this.specs) {
+                    this.processSharedSpec(
+                        compilation, allChunks, globalModulesSet, spec);
+                }
+            });
+
+            // additional-chunk-assets is the only unconditional hook that gets
+            // passed all of the chunks that is called after sortItemsWithModuleIds
+            // and sortItemsWithChunkIds which are two internal webpack methods
+            // in Compilation.js that sort the order of the parents based on
+            // chunk id.
+            compilation.plugin(["additional-chunk-assets"], (allChunks) => {
```

Comment:
Interesting - even though these two plugin calls have callbacks they're being executing synchronously?

Reply:
There are a bunch of different lifecycle hooks that a plugin can register for and they're called in a specific sequence.  This one gets called after `"optimize-chunks"`.  I'll link to the documentation and call out the ordering specifically.